_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchts2
+date: 03-15-2021
+categories:
+  - Torch
+  - R
+  - Time Series
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Today, we continue our exploration of multi-step time-series forecasting with `torch`. This post is the third in a series.
+
+-   [Initially](https://blogs.rstudio.com/ai/posts/2021-03-10-forecasting-time-series-with-torch_1/), we covered basics of recurrent neural networks (RNNs), and trained a model to predict the very next value in a sequence. We also found we could forecast quite a few steps ahead by feeding back individual predictions in a loop.
+
+-   [Next](https://blogs.rstudio.com/ai/posts/2021-03-11-forecasting-time-series-with-torch_2/), we built a model "natively" for multi-step prediction. A small multi-layer-perceptron (MLP) was used to project RNN output to several time points in the future.
+
+Of both approaches, the latter was the more successful. But conceptually, it has an unsatisfying touch to it: When the MLP extrapolates and generates output for, say, ten consecutive points in time, there is no causal relation between those. (Imagine a weather forecast for ten days that never got updated.)
+
+Now, we'd like to try something more intuitively appealing. The input is a sequence; the output is a sequence. In natural language processing (NLP), this type of task is very common: It's exactly the kind of situation we see with machine translation or summarization.
+
+Quite fittingly, the types of models employed to these ends are named sequence-to-sequence models (often abbreviated *seq2seq*). In a nutshell, they split up the task into two components: an encoding and a decoding part. The former is done just once per input-target pair. The latter is done in a loop, as in our first try. But the decoder has more information at its disposal: At each iteration, its processing is based on the previous prediction as well as previous state. That previous state will be the encoder's when a loop is started, and its own ever thereafter.
+
+Before discussing the model in detail, we need to adapt our data input mechanism.
+
+# Data input
+
+We continue working with `vic_elec` , provided by `tsibbledata`.
+
+Again, the dataset definition in the current post looks a bit different from the way it did before; again, it's the shape of the target that differs. This time, `y` equals `x`, shifted to the left by one.
+
+The reason we do this is owed to the way we are going to train the network. With *seq2seq*, people often use a technique called "teacher forcing", where instead of feeding back its own prediction into the decoder module, you pass it the value it *should* have predicted. To be clear, this is done during training only, and to a configurable degree.
```

Comment:
```suggestion
The reason we do this is owed to the way we are going to train the network. With *seq2seq*, people often use a technique called "teacher forcing" where, instead of feeding back its own prediction into the decoder module, you pass it the value it *should* have predicted. To be clear, this is done during training only, and to a configurable degree.
```